Content.Server/Mining/Electromagnet.cs

```diff
@@ -0,0 +1,79 @@
+using Content.Server.Power.Components;
+using Content.Shared.Tag;
+using Robust.Shared.Physics.Components;
+using Robust.Shared.Physics.Systems;
+
+namespace Content.Server.Mining;
+
+[RegisterComponent]
+public class ElectromagnetComponent : Component
+{
+    [DataField("range")]
+    [ViewVariables(VVAccess.ReadWrite)]
+    public float Range = 2f;
+
+    [DataField("force")]
+    [ViewVariables(VVAccess.ReadWrite)]
+    public float Force = 300f;
+
+    [DataField("minDist")]
+    [ViewVariables(VVAccess.ReadWrite)]
+    public float MinDist = 0.5f;
+}
+
+public class ElectromagnetSystem : EntitySystem
+{
+    [Dependency] private readonly EntityLookupSystem _lookup = default!;
+    [Dependency] private readonly TagSystem _tagSystem = default!;
+    [Dependency] private readonly SharedPhysicsSystem _physics = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+    }
+
+    public override void Update(float frameTime)
+    {
+        foreach (var comp in EntityManager.EntityQuery<ElectromagnetComponent>())
+        {
+            if (!(TryComp<ApcPowerReceiverComponent>(comp.Owner, out var apc) && apc.Powered))
+            {
+                // Not powered, don't do anything
+                continue;
+            }
+
+            var xformQuery = GetEntityQuery<TransformComponent>();
+            var physQuery = GetEntityQuery<PhysicsComponent>();
+            if (!xformQuery.TryGetComponent(comp.Owner, out var myXform))
+            {
+                // We need our own transform in order to move things
+                continue;
+            }
+
+            foreach (var uid in _lookup.GetEntitiesInRange(comp.Owner, comp.Range))
+            {
+                if (_tagSystem.HasTag(uid, "Metal"))
+                {
+                    if (!xformQuery.TryGetComponent(uid, out var xform))
+                        continue;
+
+                    if (!physQuery.TryGetComponent(uid, out var phys))
+                        continue;
+
+                    var dx = myXform.WorldPosition - xform.WorldPosition;
+                    var dir = dx.Normalized;
+                    var dist = dx.Length;
+                    if (dist > comp.MinDist)
+                    {
+                        var force = dir * (comp.Force / MathF.Pow(dist, 2)); // falls off inverse squared wrt distance
+                        _physics.ApplyForce(uid, force);
+                    }
+                    else
+                    {
+                        _physics.SetLinearVelocity(uid, Vector2.Zero);
+                    }
+                }
+            }
+        }
```

Comment:
Should this just use gravitywell? It's also going to be very expensive, have you profiled it?

It's also using the obsolete worldposition methods.

Reply:
No, I cannot "just" use gravitywell, because that would pull all entities, not just those with a `Metal` tag.

No, I haven't profiled this. Which part of this do you expect to be expensive?

Whatever expensive is offset by the default range of this entity lookup being 2, and that there are not that many electromagnets running about.

Reply:
> No, I haven't profiled this. Which part of this do you expect to be expensive?
World position checks, especially without using EntityQuery<T> more.

> No, I cannot "just" use gravitywell, because that would pull all entities

An event won't work? CanGravPulse just needs converting to use one and it seems redundant to maintain 2 separate pieces of gravity code.

> Whatever expensive is offset by the default range of this entity lookup being 2

Singularity is singular but is off-station, whereas this can be dragged around to areas such as the bar where there's hundreds of entities.

Reply:
But this isn't a gravity pulse! Among other things:

- The force doesn't scale with mass.
- This is continuous, and not pulsed.
- This needs to force the velocity to zero (not just zero the force) to create the illusion that the object has snapped onto the magnet, much like how you attach a magnet to your refrigerator; once it snaps it stops.

I understand the desire to duplicate code. It is nice to be able to write "general" code once and be able to apply it to other places. But with all due respect, I think here the behavior is different enough to warrant a separate entity system and component. The Electromagnet code is dead simple; there's no need to shove a triangle into a square hole.